### DISCUSSION TITLE
Dbus message_bus.py error on line 683 [Errno 2] No such file or directory

### DISCUSSION THREAD

**Discussion by FormanLukas** [hbldh/bleak#1421] [Q&A]
Hello everyone,

I am a python newbie, and maybe its too much for me. I am trying to write my program for Homeassistant to integrate LEGO hub using pybrics firmware and bleak. But cannot make any progress with any code I try.

Version of Bleak: 0.20.2
Version of Dbus-fast: 1.94.0
HW: Odroid N2+
USB Bluetooth: ODROID Wi-Fi modul 5BK (Realtek RTL8821CU)

Everytime every code I try will end by error on line 683 [Errno 2] No such file or directory in Dbus message_bus.py.

For example I tryied this code (and many more). I know that at the moment there are probably many more mistakes more far in the code, but it is still ends even in the Bleak part.

```python
# SPDX-License-Identifier: MIT
# Copyright (c) 2020 Henrik Blidh
# Copyright (c) 2022 The Pybricks Authors

import asyncio
from bleak import BleakScanner, BleakClient

UART_SERVICE_UUID = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E"
UART_RX_CHAR_UUID = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E"
UART_TX_CHAR_UUID = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E"

# Replace this with the name of your hub if you changed
# it when installing the Pybricks firmware.
HUB_NAME = "Parni_masinka"


def hub_filter(device, ad):
    return device.name and device.name.lower() == HUB_NAME.lower()


def handle_disconnect(_):
    print("Hub was disconnected.")


def handle_rx(_, data: bytearray):
    print("Received:", data)


async def main():
    # Find the device and initialize client.
    device = await BleakScanner.find_device_by_filter(hub_filter)
    client = BleakClient(device, disconnected_callback=handle_disconnect)

    # Shorthand for sending some data to the hub.
    async def send(client, data):
        await client.write_gatt_char(rx_char, data)

    try:
        # Connect and get services.
        await client.connect()
        await client.start_notify(UART_TX_CHAR_UUID, handle_rx)
        nus = client.services.get_service(UART_SERVICE_UUID)
        rx_char = nus.get_characteristic(UART_RX_CHAR_UUID)

        # Tell user to start program on the hub.
        print("Start the program on the hub now with the button.")

        # Send a few messages to the hub.
        for i in range(5):
            await send(client, b"fwd")
            await asyncio.sleep(1)
            await send(client, b"rev")
            await asyncio.sleep(1)

        # Send a message to indicate stop.
        await send(client, b"bye")

    except Exception as e:
        # Handle exceptions.
        print(e)
    finally:
        # Disconnect when we are done.
        await client.disconnect()


# Run the main async program.
asyncio.run(main())
```
Or I used this one but with the same result:
```python

# SPDX-License-Identifier: MIT
# Copyright (c) 2020 Henrik Blidh
# Copyright (c) 2022 The Pybricks Authors

"""
UART Service
-------------

An example showing how to write a simple program using the Nordic Semiconductor
(nRF) UART service.

"""

import asyncio
import sys
from itertools import count, takewhile
from typing import Iterator

from bleak import BleakClient, BleakScanner
from bleak.backends.characteristic import BleakGATTCharacteristic
from bleak.backends.device import BLEDevice
from bleak.backends.scanner import AdvertisementData

UART_SERVICE_UUID = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E"
UART_RX_CHAR_UUID = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E"
UART_TX_CHAR_UUID = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E"


# TIP: you can get this function and more from the ``more-itertools`` package.
def sliced(data: bytes, n: int) -> Iterator[bytes]:
    return takewhile(len, (data[i : i + n] for i in count(0, n)))


async def uart_terminal():
    """This is a simple "terminal" program that uses the Nordic Semiconductor
    (nRF) UART service. It reads from stdin and sends each line of data to the
    remote device. Any data received from the device is printed to stdout.
    """

    def match_nus_uuid(device: BLEDevice, adv: AdvertisementData):
        # This assumes that the device includes the UART service UUID in the
        # advertising data. This test may need to be adjusted depending on the
        # actual advertising data supplied by the device.
        if UART_SERVICE_UUID.lower() in adv.service_uuids:
            return True

        return False

    device = await BleakScanner.find_device_by_filter(match_nus_uuid)

    if device is None:
        print("no matching device found, you may need to edit match_nus_uuid().")
        sys.exit(1)

    def handle_disconnect(_: BleakClient):
        print("Device was disconnected, goodbye.")
        # cancelling all tasks effectively ends the program
        for task in asyncio.all_tasks():
            task.cancel()

    def handle_rx(_: BleakGATTCharacteristic, data: bytearray):
        print("received:", data)

    async with BleakClient(device, disconnected_callback=handle_disconnect) as client:
        await client.start_notify(UART_TX_CHAR_UUID, handle_rx)

        print("Connected, start typing and press ENTER...")

        loop = asyncio.get_running_loop()
        nus = client.services.get_service(UART_SERVICE_UUID)
        rx_char = nus.get_characteristic(UART_RX_CHAR_UUID)

        while True:
            # This waits until you type a line and press ENTER.
            # A real terminal program might put stdin in raw mode so that things
            # like CTRL+C get passed to the remote device.
            data = await loop.run_in_executor(None, sys.stdin.buffer.readline)

            # data will be empty on EOF (e.g. CTRL+D on *nix)
            if not data:
                break

            # some devices, like devices running MicroPython, expect Windows
            # line endings (uncomment line below if needed)
            # data = data.replace(b"\n", b"\r\n")

            # Writing without response requires that the data can fit in a
            # single BLE packet. We can use the max_write_without_response_size
            # property to split the data into chunks that will fit.

            for s in sliced(data, rx_char.max_write_without_response_size):
                await client.write_gatt_char(rx_char, s, response=False)

            print("sent:", data)

# Run the main async program.
asyncio.run(uart_terminal())
```
Please be tolerant as I am new one and trying hard project. I will be glad for any advice. 

Thanks

**Comment by dlech** [comment]
Can you share the full stack trace of the error? It sounds like the dbus daemon may not be running.

**Reply by FormanLukas** [reply]
Yes, here it is:

➜  / /bin/python3 /config/custom_components/pybrics/pokus_1.py
Traceback (most recent call last):
  File "/config/custom_components/pybrics/pokus_1.py", line 67, in <module>
    asyncio.run(main())
  File "/usr/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/config/custom_components/pybrics/pokus_1.py", line 31, in main
    device = await BleakScanner.find_device_by_filter(hub_filter)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/bleak/__init__.py", line 441, in find_device_by_filter
    async with cls(**kwargs) as scanner:
  File "/usr/local/lib/python3.11/dist-packages/bleak/__init__.py", line 158, in __aenter__
    await self._backend.start()
  File "/usr/local/lib/python3.11/dist-packages/bleak/backends/bluezdbus/scanner.py", line 168, in start
    manager = await get_global_bluez_manager()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/bleak/backends/bluezdbus/manager.py", line 1046, in get_global_bluez_manager
    await instance.async_init()
  File "/usr/local/lib/python3.11/dist-packages/bleak/backends/bluezdbus/manager.py", line 235, in async_init
    bus = MessageBus(bus_type=BusType.SYSTEM, auth=get_dbus_authenticator())
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/dbus_fast/aio/message_bus.py", line 198, in __init__
    super().__init__(bus_address, bus_type, ProxyObject, negotiate_unix_fd)
  File "/usr/local/lib/python3.11/dist-packages/dbus_fast/message_bus.py", line 161, in __init__
    self._setup_socket()
  File "/usr/local/lib/python3.11/dist-packages/dbus_fast/message_bus.py", line 714, in _setup_socket
    raise err
  File "/usr/local/lib/python3.11/dist-packages/dbus_fast/message_bus.py", line 687, in _setup_socket
    self._sock.connect(filename)
FileNotFoundError: [Errno 2] No such file or directory

**Reply by dlech** [reply]
Which D-Bus daemon are you using?

**Reply by FormanLukas** [reply]
I dont know how to find out in Home Assistant. I thought, that there is some, because other (already supported BLE devices works). I think that the issue is caused by something else. Other instaled integration for some BLE devices which programed anyone else do the same error if I just run the relevant python file in visual studio code. So I probably making something totaly wrong and the right issue is anywhere else (in my missing knowledge). Maybe it is too much for me - I am totaly lost.

**Reply by xpj** [reply]
Observing the same behavior with the simple scanning example (https://github.com/hbldh/bleak/blob/develop/examples/scan_iterator.py), but when running in Podman container. It could be caused by non-root container and not mapped root filesystem.

Fedora 38 on Raspberry PI 4b
Bus 001 Device 004: ID 0a12:0001 Cambridge Silicon Radio, Ltd Bluetooth Dongle (HCI mode)

D-Bus Message Bus Daemon 1.14.6
bleak-0.21.1
dbus_fast-2.12.0

```
Traceback (most recent call last):
  File "/main.py", line 22, in <module>
    asyncio.run(main())
  File "/usr/local/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/main.py", line 10, in main
    async with BleakScanner() as scanner:
  File "/usr/local/lib/python3.10/site-packages/bleak/__init__.py", line 158, in __aenter__
    await self._backend.start()
  File "/usr/local/lib/python3.10/site-packages/bleak/backends/bluezdbus/scanner.py", line 168, in start
    manager = await get_global_bluez_manager()
  File "/usr/local/lib/python3.10/site-packages/bleak/backends/bluezdbus/manager.py", line 1046, in get_global_bluez_manager
    await instance.async_init()
  File "/usr/local/lib/python3.10/site-packages/bleak/backends/bluezdbus/manager.py", line 235, in async_init
    bus = MessageBus(bus_type=BusType.SYSTEM, auth=get_dbus_authenticator())
  File "/usr/local/lib/python3.10/site-packages/dbus_fast/aio/message_bus.py", line 198, in __init__
    super().__init__(bus_address, bus_type, ProxyObject, negotiate_unix_fd)
  File "/usr/local/lib/python3.10/site-packages/dbus_fast/message_bus.py", line 161, in __init__
    self._setup_socket()
  File "/usr/local/lib/python3.10/site-packages/dbus_fast/message_bus.py", line 714, in _setup_socket
    raise err
  File "/usr/local/lib/python3.10/site-packages/dbus_fast/message_bus.py", line 687, in _setup_socket
    self._sock.connect(filename)
FileNotFoundError: [Errno 2] No such file or directory
```

DBus brooker is running.
